### PR TITLE
Add support for serializing stdClass objects 

### LINKED ISSUE #33047: [Serializer] Normalizers can't handle \stdClass
**Symfony version(s) affected**: 4.3.3

**Description**  
When an object contains properties of `\stdClass`, serialization fails.

[Possible secondary issue: `ObjectNormalizer` fails to read data from this object]

**How to reproduce**  
My first attempt with the standard Symfony configuration of the serializer (using `ObjectNormalizer` is missing most properties.

*Code*
```
<?php
// Setup instructions:
// composer require gocardless/gocardless-pro
// composer require symfony/serializer-pack
use GoCardlessPro\Resources\Event;
use Symfony\Component\Serializer\Encoder\JsonEncoder;
use Symfony\Component\Serializer\Encoder\XmlEncoder;
use Symfony\Component\Serializer\Normalizer\ObjectNormalizer;
use Symfony\Component\Serializer\Serializer;

require_once 'vendor/autoload.php';

// This - I hope - mimics the Symfony framework default Serializer configuration
$encoders = [new XmlEncoder(), new JsonEncoder()];
$normalizers = [new ObjectNormalizer()];
$serializer = new Serializer($normalizers, $encoders);
// End configuring the serializer

// Begin configuring the object to serialize
$json = '{"id":"EV0000000000000","created_at":"2019-08-06T20:09:15.731Z","action":"active","resource_type":"mandates","details":{"origin":"gocardless","cause":"mandate_activated","description":"The time window after submission for the banks to refuse a mandate has ended without any errors being received, so this mandate is now active."},"links":{"mandate":"MD00000000000"},"metadata":[],"model_name":"Event"}';
$event = new Event(json_decode($json));

echo "Expected\n";
// Get rid of the line breaks:
echo json_encode(json_decode($json));

echo "\n\nActual\n";
echo $serializer->serialize($event, 'json');
```
*Output*
```
Expected
{"id":"EV0000000000000","created_at":"2019-08-06T20:09:15.731Z","action":"active","resource_type":"mandates","details":{"origin":"gocardless","cause":"mandate_activated","description":"The time window after submission for the banks to refuse a mandate has ended without any errors being received, so this mandate is now active."},"links":{"mandate":"MD00000000000"},"metadata":[],"model_name":"Event"}

Actual
{"api_response":null}
```

I assume that is expected behaviour (though don't understand why) so tried a different normalizer: `PropertyNormalizer`. This worked if the structure was all `array`s (see comment in code) but not with objects:

*Code*
```
<?php
// Setup instructions:
// composer require gocardless/gocardless-pro
// composer require symfony/serializer-pack
use GoCardlessPro\Resources\Event;
use Symfony\Component\Serializer\Encoder\JsonEncoder;
use Symfony\Component\Serializer\Encoder\XmlEncoder;
use Symfony\Component\Serializer\Normalizer\PropertyNormalizer;
use Symfony\Component\Serializer\Serializer;

require_once 'vendor/autoload.php';

// Custom Serializer configuration
$encoders = [new XmlEncoder(), new JsonEncoder()];
$normalizers = [new PropertyNormalizer()];
$serializer = new Serializer($normalizers, $encoders);
// End configuring the serializer

$json = '{"id":"EV0000000000000","created_at":"2019-08-06T20:09:15.731Z","action":"active","resource_type":"mandates","details":{"origin":"gocardless","cause":"mandate_activated","description":"The time window after submission for the banks to refuse a mandate has ended without any errors being received, so this mandate is now active."},"links":{"mandate":"MD00000000000"},"metadata":[],"model_name":"Event"}';
// All problems can be fixed if I could change this to $event = new Event(json_decode($json, true));
// But I can't - the library that creates these objects doesn't
// So I need the serializer to handle stdClass objects
$event = new Event(json_decode($json));

echo "Expected\n";
// Get rid of the line breaks:
echo json_encode(json_decode($json));

echo "\n\nActual\n";
// IMO this should work, but it throws an exception:
// PHP Fatal error:  Uncaught Symfony\Component\Serializer\Exception\NotNormalizableValueException: Could not normalize object of type stdClass, no supporting normalizer found.
#echo $serializer->serialize($event, 'json');

// So I try doing my own conversions:
// It still throws the same exception
$callback = function ($innerObject, $outerObject, string $attributeName, string $format = null, array $context = []) {
    return $innerObject instanceof \stdClass ? (array)$innerObject : $innerObject;
};
echo $serializer->serialize($event, 'json', [
    'callbacks' => [
        // This assumes I know all the object properties that are stdClass - which I won't
        'details' => $callback,
        'links' => $callback,
        'metadata' => $callback,
        'data' => $callback,
    ]
]);
```

*Output*
```
Expected
{"id":"EV0000000000000","created_at":"2019-08-06T20:09:15.731Z","action":"active","resource_type":"mandates","details":{"origin":"gocardless","cause":"mandate_activated","description":"The time window after submission for the banks to refuse a mandate has ended without any errors being received, so this mandate is now active."},"links":{"mandate":"MD00000000000"},"metadata":[],"model_name":"Event"}

Actual
PHP Fatal error:  Uncaught Symfony\Component\Serializer\Exception\NotNormalizableValueException: Could not normalize object of type stdClass, no supporting normalizer found. in Z:\path\to\vendor\symfony\serializer\Serializer.php:173
Stack trace:
#0 Z:\path\to\vendor\symfony\serializer\Serializer.php(162): Symfony\Component\Serializer\Serializer->normalize(Object(stdClass), 'json', Array)
#1 Z:\path\to\vendor\symfony\serializer\Normalizer\AbstractObjectNormalizer.php(206): Symfony\Component\Serializer\Serializer->normalize(Array, 'json', Array)
#2 Z:\path\to\vendor\symfony\serializer\Serializer.php(152): Symfony\Component\Serializer\Normalizer\AbstractObjectNormalizer->normalize(Object(GoCardlessPro\Resources\Event), 'json', Array)
#3 Z:\path\to\vendor\symfony\serializer\Serializer.php(125): Symfony\Component\Serializer\Serializer->normalize in Z:\path\to\vendor\symfony\serializer\Serializer.php on line 173

Fatal error: Uncaught Symfony\Component\Serializer\Exception\NotNormalizableValueException: Could not normalize object of type stdClass, no supporting normalizer found. in Z:\path\to\vendor\symfony\serializer\Serializer.php on line 173

Symfony\Component\Serializer\Exception\NotNormalizableValueException: Could not normalize object of type stdClass, no supporting normalizer found. in Z:\path\to\vendor\symfony\serializer\Serializer.php on line 173

Call Stack:
    0.0002     405288   1. {main}() Z:\path\to\index.php:0
    0.0130    1260760   2. Symfony\Component\Serializer\Serializer->serialize() Z:\path\to\index.php:45
    0.0131    1261136   3. Symfony\Component\Serializer\Serializer->normalize() Z:\path\to\vendor\symfony\serializer\Serializer.php:125
    0.0131    1262264   4. Symfony\Component\Serializer\Normalizer\PropertyNormalizer->normalize() Z:\path\to\vendor\symfony\serializer\Serializer.php:152
    0.0138    1267416   5. Symfony\Component\Serializer\Serializer->normalize() Z:\path\to\vendor\symfony\serializer\Normalizer\AbstractObjectNormalizer.php:206
    0.0138    1267792   6. Symfony\Component\Serializer\Serializer->normalize() Z:\path\to\vendor\symfony\serializer\Serializer.php:162
```

**Possible Solution**  
Write ones own normalizer for `stdClass`? As it's a PHP built-in class, I'm not clear this is intended; why can't the standard object normalizers cope with it?

**Additional context**

### PR DESCRIPTION
**Pull request by henry2778**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #33047
| License       | MIT
| Doc PR        | todo

Updated PR for issue #33047 :) Adding support for serializing stdClass objects, because it was missing. Thanks!

### CODE REVIEW

File: src/Symfony/Component/Serializer/Tests/SerializerTest.php [hunk 1]
```diff
@@ -81,7 +82,7 @@ public function testNormalizeNoMatch()
     {
         $this->expectException('Symfony\Component\Serializer\Exception\UnexpectedValueException');
         $serializer = new Serializer([$this->getMockBuilder('Symfony\Component\Serializer\Normalizer\CustomNormalizer')->getMock()]);
-        $serializer->normalize(new \stdClass(), 'xml');
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/SerializerTest.php]: This change looks like a tiny BC break, on the other hand this test was relying on an (undocumented) edge case. WDYT @symfony/mergers?

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Component/Serializer/Tests/SerializerTest.php (hunk 1 above)
Comment (weaverryan) [reply src/Symfony/Component/Serializer/Tests/SerializerTest.php]: What’s the change? That previously you would get an exception with stdClass and now it would work? Or am I missing the point?

Comment (henry2778) [reply src/Symfony/Component/Serializer/Tests/SerializerTest.php]: I guess that test covers the case when we have no matching normalizer for the class that is normalizable. So its not really important which not-normalizable class we are trying to normalize

Comment (chalasr) [reply src/Symfony/Component/Serializer/Tests/SerializerTest.php]: I think what @dunglas means is that currently an instance of `stdClass` can be properly serialized if you register a custom normalizer that supports it, with this PR such normalizer wouldn't be called anymore. 
IMHO it's ok but only if we don't match classes extending `stdClass` (yes, one can do that..), I suggest using `\stdClass::class === \get_class($data)` instead of `instanceof`.

**Review by chalasr — CHANGES_REQUESTED** [review]
See my comment above

File: src/Symfony/Component/Serializer/Tests/SerializerTest.php (hunk 1 above)
Comment (ogizanagi) [reply src/Symfony/Component/Serializer/Tests/SerializerTest.php]: @chalasr : the ObjectNormalizer has a [very low priority](https://github.com/symfony/symfony/blob/e3b513b5a154d0a0f5e97a27bfd8108084f0a0df/src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.xml#L71-L72) though. So I don't think that's an issue.

Comment (ogizanagi) [reply src/Symfony/Component/Serializer/Tests/SerializerTest.php]: Why not making it part of the object normalizer instead? It has a very low priority that would mitigate this issue.

Comment (henry2778) [reply src/Symfony/Component/Serializer/Tests/SerializerTest.php]: @ogizanagi thanks, good idea. Will check that tomorrow, I think that will work

**Comment by henry2778** [comment]
@dunglas @chalasr please have a look at my implementation of @ogizanagi suggestion. Thanks!

**Comment by yoshz** [comment]
Any update on this MR? This is a quite annoying bug.

**Comment by fabpot** [comment]
ping @dunglas

**Comment by dunglas** [comment]
The new implementation doesn't look good to me. It doesn't support any of the feature of `ObjectNormalizer` (the circular references and max depth handler, the callbacks...), while it should. Maybe than extracting `stdClass`'s properties could be handled in `extractAttributes` instead.

**Comment by dunglas** [comment]
I just opened an alternative PR with what I've in mind: https://github.com/symfony/symfony/pull/35596

**Comment by chalasr** [comment]
Thank you for pushing this @henry2778.
Closing in favor of #35596
